core/src/audio_subsystem.cpp

```diff
@@ -0,0 +1,206 @@
+#include <knoting/audio_subsystem.h>
+
+#include <knoting/audio_source.h>
+
+namespace knot {
+
+AudioSubsystem::AudioSubsystem() = default;
+
+void AudioSubsystem::on_awake() {
+    void* extra_driver_data = nullptr;
+    m_result = System_Create(&m_system);
+    ErrorCheck(m_result);
+    m_result = m_system->init(512, FMOD_INIT_NORMAL, extra_driver_data);
+    ErrorCheck(m_result);
+    m_result = m_system->set3DSettings(1.0f, 1.0f, 1.0f);
+    ErrorCheck(m_result);
+    log::debug("AudioSubsystem created!");
+}
+
+void AudioSubsystem::on_destroy() {
+    m_result = m_system->close();
+    ErrorCheck(m_result);
+    m_result = m_system->release();
+    ErrorCheck(m_result);
+}
+
+void AudioSubsystem::play(components::AudioSource& source) {
+    source.set_playing(false);
+    if (source.get_channel()) {
+        m_result = source.get_channel()->isPlaying(&source.get_playing());
+        ErrorCheck(m_result);
+    }
+
+    if (!source.get_playing()) {
+        m_result = m_system->playSound(source.get_sound(), nullptr, false, &source.get_channel());
+        ErrorCheck(m_result);
+
+        update_source(source);
+
+        source.set_playing(true);
+    }
+}
+
+void AudioSubsystem::toggle(components::AudioSource& source) {
+    if (!source.get_channel())
+        return;
+
+    source.get_paused() = !source.get_paused();
+    source.get_channel()->setPaused(source.get_paused());
+}
+
+void AudioSubsystem::stop(components::AudioSource& source) {
+    if (!source.get_channel())
+        return;
+
+    source.set_playing(false);
+    source.get_channel()->setPaused(true);
+    source.set_channel(nullptr);
+}
+
+void AudioSubsystem::stop() {
+    auto sceneOpt = Scene::get_active_scene();
+    if (!sceneOpt)
+        return;
+
+    auto scene = sceneOpt.value();
+    auto view = scene.get().get_registry().view<components::AudioSource>();
```

Comment:
could we give these more meaningful names?
view could be sources and "entt" can be source